src/h3/frame.rs

```diff
@@ -0,0 +1,856 @@
+    let mut num_placeholders = None;
+    let mut max_header_list_size = None;
+    let mut qpack_max_table_capacity = None;
+    let mut qpack_blocked_streams = None;
+
+    while b.off() < payload_length as usize {
+        // TODO test this exit condition
+        let setting = b.get_u16()?;
+
+        match setting {
+            0x1 => {
+                qpack_max_table_capacity = Some(b.get_varint()?);
+            },
+            0x6 => {
+                max_header_list_size = Some(b.get_varint()?);
+            },
+            0x7 => {
+                qpack_blocked_streams = Some(b.get_varint()?);
+            },
+            0x8 => {
+                num_placeholders = Some(b.get_varint()?);
+            },
+            _ => {
+                // TODO: not implemented
+            },
+        }
+    }
+
+    Ok(H3Frame::Settings {
+        num_placeholders,
+        max_header_list_size,
+        qpack_max_table_capacity,
+        qpack_blocked_streams,
+    })
+}
+
+fn parse_priority_frame(b: &mut octets::Octets) -> Result<H3Frame> {
+    let bitfield = b.get_u8()?;
+    let mut prioritized_element_id = None;
+    let mut element_dependency_id = None;
+
+    let priority_elem = PrioritizedElemType::from_bits(bitfield >> 6);
+
+    let elem_dependency = ElemDependencyType::from_bits(
+        (bitfield & ELEM_DEPENDENCY_TYPE_MASK) >> 4,
+    );
+
+    if !priority_elem.is_peid_absent() {
+        prioritized_element_id = Some(b.get_varint()?);
+    }
+
+    if !elem_dependency.is_edid_absent() {
+        element_dependency_id = Some(b.get_varint()?);
+    }
+
+    let weight = b.get_u8()?;
+    Ok(H3Frame::Priority {
+        priority_elem,
+        elem_dependency,
+        prioritized_element_id,
+        element_dependency_id,
+        weight,
+    })
+}
+
+fn parse_push_promise(
+    payload_length: u64, b: &mut octets::Octets,
+) -> Result<H3Frame> {
+    let push_id = b.get_varint()?;
+    let header_block_length = payload_length - octets::varint_len(push_id) as u64;
+    let header_block = b.get_bytes(header_block_length as usize)?.to_vec();
+
+    Ok(H3Frame::PushPromise {
+        push_id,
+        header_block,
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn data() {
+        let mut d: [u8; 128] = [42; 128];
+
+        let frame = H3Frame::Data {
+            payload: vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
+        };
+
+        let wire_len = {
+            let mut b = octets::Octets::with_slice(&mut d);
+            frame.to_bytes(&mut b).unwrap()
+        };
+
+        assert_eq!(wire_len, 14);
+
+        {
+            let mut b = octets::Octets::with_slice(&mut d);
+            assert_eq!(H3Frame::from_bytes(&mut b).unwrap(), frame);
+        }
+    }
+
+    #[test]
+    fn headers() {
+        let mut d: [u8; 128] = [42; 128];
+
+        // TODO test QPACK'd headers
+        let frame = H3Frame::Headers {
+            header_block: vec![71, 69, 84, 32, 47],
```

Comment:
Nit: for ASCII constants it's better to use a bytestring literal for readability. E.g. here it would be `b"GET /".to_vec()`.

Reply:
You've drawn my attention to a TODO that really needs addressing. `header_block` should contain a set of QPACK instructions but these tests were written before we had a QPACK codec. 

Right now, the tests create HEADERS and PUSH_PROMISE frames containing garbage values (as far as a strict protocol check would determine), which just so happen to also be an ASCII string that has a striking resemblance to HTTP/1.x syntax. 

I'm inclined to say that the unit test should not use QPACK nor something that can be interpreted as an ASCII string. Thoughts @ghedo?